dcc-portal-ui/app/scripts/survivalanalysis/js/service.js

```diff
@@ -0,0 +1,199 @@
+          .join('\n');
+
+        var tsv = headings + '\n' + contents;
+        return tsv;
+      }
+
+      _.extend(this, {
+        fetchSurvivalData: fetchSurvivalData,
+        dataSetToTSV: dataSetToTSV
+      });
+
+    });
+
+  module.service('SurvivalAnalysisLaunchService', function(SetService, Restangular, LocationService,
+    SetNameService, $timeout, $location, Page){
+      const _service = this;
+
+      var d1, d2, type = 'donor';
+
+      _service.getSetName = function(filters){
+        return SetNameService.getSetFilters()
+          .then(function (filters) {
+            return SetNameService.getSetName(filters);
+          });
+      }
+
+      // Wait for sets to materialize
+      _service.wait = (ids, numTries, callback) => {
+        if (numTries <= 0) {
+          return;
+        }
+        SetService.getMetaData(ids).then(function(data) {
+          var finished = _.filter(data, function(d) {
+            return d.state === 'FINISHED';
+          });
+
+          if (finished.length === ids.length) {
+            callback(data);
+          } else {
+            $timeout(function() {
+              _service.wait(ids, --numTries, callback);
+            }, 1500);
+          }
+        });
+      }
+
+       /* Phenotype comparison only takes in donor set ids */
+      _service.launchAnalysis = (setIds) => {
+        return Restangular
+          .one('analysis')
+          .post('phenotype', setIds, {}, {'Content-Type': 'application/json'})
+          .then(function(data) {
+            if (!data.id) {
+              console.log('Could not retrieve analysis data.id');
+            }
+            LocationService.goToPath('analysis/view/phenotype/' + data.id);
+          }).finally(function(){
+            Page.stopWork();
+          });
+      };
+
+      _service.launchSurvivalAnalysis = (entityType, entityId, entitySymbol, filters) => {
+
+        Page.startWork();
+
+        _service.getSetName(filters).then(function (setName) {
+          d1 = {
+            filters: entityType === 'gene' ? _.extend(_.clone(filters), {gene: { id: { is: [entityId] } }}) : _.extend(_.clone(filters), {mutation: { id: { is: [entityId] } }}),
+            isTransient: true,
+            type: type,
+            name: entityType === 'gene' ? entitySymbol + ' Mutated Donors ' + setName : 'Donors with ' + entitySymbol + ' ' + setName
+          };
+
+          d2 = {
+            filters: entityType === 'gene' ? _.extend(_.clone(filters), {gene: { id: { not: [entityId] } }}) : _.extend(_.clone(filters), {mutation: { id: { not: [entityId] } }}),
+            isTransient: true,
+            type: type,
+            name: entityType === 'gene' ? entitySymbol + ' Not Mutated Donors ' + setName : 'Donors without ' + entitySymbol + ' ' + setName
+          };
+
+          var setIds = [];
+
+          SetService.addSet(type, d1).then(function (r1) {
+            setIds.push(r1.id);
+            SetService.addSet(type, d2).then(function (r2) {
```

Comment:
Yup!

Aside from nesting, the current code also has an undesired effect of having to wait until `SetService.addSet(type, d1)` returns before firing off `SetService.addSet(type, d2)`, when these are independent actions that can be fired off simultaneously

With async/await, we can now do this:

```js
const [r1, r2] = [await SetService.addSet(type, d1), await SetService.addSet(type, d2)];
setIds = setIds.concat([r1.id, r2.id]);
_service.wait(setIds, 7, () => _service.launchAnalysis(setIds));
```